scanpy/plotting/anndata.py

```diff
@@ -1657,38 +1798,363 @@ def matrixplot(adata, var_names, groupby=None, use_raw=None, log=False, num_cate
+        from matplotlib import gridspec
+        fig = pl.figure(figsize=(width, height))
+        axs = gridspec.GridSpec(nrows=2, ncols=3, wspace=0.02, hspace=0.04,
+                                width_ratios=[heatmap_width, dendro_width, colorbar_width],
+                                height_ratios=height_ratios)
 
-    # invert y axis to show categories ordered from top to bottom
-    matrix_ax.set_ylim(mean_obs.shape[0], 0)
+        matrix_ax = fig.add_subplot(axs[1, 0])
+        y_ticks = np.arange(mean_obs.shape[0]) + 0.5
+        matrix_ax.set_yticks(y_ticks)
+        matrix_ax.set_yticklabels([mean_obs.index[idx] for idx in range(mean_obs.shape[0])])
 
-    x_ticks = np.arange(mean_obs.shape[1]) + 0.5
-    matrix_ax.set_xticks(x_ticks)
-    matrix_ax.set_xticklabels([mean_obs.columns[idx] for idx in range(mean_obs.shape[1])], rotation=90)
-    matrix_ax.tick_params(axis='both', labelsize='small')
-    matrix_ax.grid(False)
-    matrix_ax.set_xlim(-0.5, len(var_names) + 0.5)
-    matrix_ax.set_ylabel(groupby)
-    matrix_ax.set_xlim(0, mean_obs.shape[1])
+        color_legend = fig.add_subplot(axs[1, 2])
 
-    # plot group legends on top of matrix_ax (if given)
-    if var_group_positions is not None and len(var_group_positions) > 0:
-        gene_groups_ax = fig.add_subplot(axs[0, 0], sharex=matrix_ax)
-        _plot_gene_groups_brackets(gene_groups_ax, group_positions=var_group_positions,
-                                   group_labels=var_group_labels, rotation=var_group_rotation,
-                                   left_adjustment=0.2, right_adjustment=0.8)
+        if dendrogram:
+            dendro_ax = fig.add_subplot(axs[1, 1], sharey=matrix_ax)
+            _plot_dendrogram(dendro_ax, adata, ticks=y_ticks)
+
+        pc = matrix_ax.pcolor(mean_obs, edgecolor='gray', **kwds)
+
+        # invert y axis to show categories ordered from top to bottom
+        matrix_ax.set_ylim(mean_obs.shape[0], 0)
+
+        x_ticks = np.arange(mean_obs.shape[1]) + 0.5
+        matrix_ax.set_xticks(x_ticks)
+        matrix_ax.set_xticklabels([mean_obs.columns[idx] for idx in range(mean_obs.shape[1])], rotation=90)
+        matrix_ax.tick_params(axis='both', labelsize='small')
+        matrix_ax.grid(False)
+        matrix_ax.set_xlim(-0.5, len(var_names) + 0.5)
+        matrix_ax.set_ylabel(groupby)
+        matrix_ax.set_xlim(0, mean_obs.shape[1])
+
+        # plot group legends on top of matrix_ax (if given)
+        if var_group_positions is not None and len(var_group_positions) > 0:
+            gene_groups_ax = fig.add_subplot(axs[0, 0], sharex=matrix_ax)
+            _plot_gene_groups_brackets(gene_groups_ax, group_positions=var_group_positions,
+                                       group_labels=var_group_labels, rotation=var_group_rotation,
+                                       left_adjustment=0.2, right_adjustment=0.8)
+
+        # plot colorbar
+        pl.colorbar(pc, cax=color_legend)
+    else:
```

Comment:
I will check if I can add some abstraction but is annoying as almost all the lines are slightly different. For example: 

![image](https://user-images.githubusercontent.com/4964309/48824325-3d8d8f00-ed64-11e8-86fc-65ffeecf2a96.png)

Reply:
I tried to reduce as much code repetition as possible. For the three branching that you pointed out it is not possible to reduce the code because 'width' and 'height' and 'x' and 'y' need to be changed in almost all lines even though most of the code remains the same.

Reply:
Oh, that’s easy! sorry, I should have been clearer on how I imagined it.

you can just name the dimensions after what they contain, and assign width and height accordingly:

```py
def heatmap(...):
    ...
    dendro_size = 1 if dendrogram else 0
    ...
    if not swap_axes:
        width = heatmap_size + dendro_size + groupby_size
```

Of course I saw that you fiddled with those, and the dimensions are different depending on the elements being horizontal or vertical. you could fix that similar to this:

```py
class Sizes(NamedTuple):
    h: float
    v: float

sizes0 = Sizes(0., 0.)
SIZES = dict(
    dendrogram=Size(h=1, v=.8),
    ....
)

def heatmap(...):
    ...
    dendro_sizes = SIZES['dendrogram'] if dendrogram else sizes0
    ...
    if not swap_axes:
        width = heatmap_sizes.h + dendro_sizes.h + groupby_sizes.h
```

Reply:
I like the idea of using a named tuple. I think I could reduce some code that way and this will also help to keep the dimensions in one place. 

However, I think that trying to find generalizable patterns for each semi-duplicated line will unnecessarily increase the complexity of the code. For example, in one case the code is: 
```PYTHON
        heatmap_ax.set_ylim(obs_tidy.shape[0] - 0.5, -0.5)
        heatmap_ax.set_xlim(-0.5, obs_tidy.shape[1] - 0.5)
        heatmap_ax.tick_params(axis='y', left=False, labelleft=False)
```
while for the other case is:
```PYTHON
        heatmap_ax.set_xlim(0, obs_tidy.shape[0])
        heatmap_ax.set_ylim(obs_tidy.shape[1] - 0.5, -0.5)
        heatmap_ax.tick_params(axis='x', bottom=False, labelbottom=False)
```

Furthermore, the the layout also changes. For example, for heatmap showing cells x genes, the the groupby categories are on the left, dendrogram and colorbar are on the right and the 'brackets' are on top (2x4 layout). For the genes x cells case, the group by categories are at the bottom, the dendrogram is at the top and 'brackets' and colorbar are to the right (3x3 layout). 

At the moment I will leave any further code improvements for the future.

Reply:
Welp, future=never, but I understand why you decided against it in the shown case.

Are you sure it applies in all cases or are there parts that can be simplified?

Reply:
Honestly, I like the SIZES idea and also the type annotations and I hope to get back to that at some point, but I need to move on to other pressing issues at the moment. 

I thought that I could not simplify any further but you showed me that it is indeed possible, so to answer your question: I don't see how to simplify further but a keen eye may find further improvements.

Reply:
OK, then! LGTM